control/HTTP.php

```diff
@@ -318,140 +319,117 @@ public static function register_etag($etag) {
+			// (http://support.microsoft.com/kb/323308)
+			// Note: this is also fixable by ticking "Do not save encrypted pages to disk" in advanced options.
+			HTTPCacheControl::inst()
+				->privateCache()
+				->removeDirective('no-cache')
+				->removeDirective('no-store');
 		}
 
-		$responseHeaders['Cache-Control'] = implode(', ', $cacheControlHeaders);
-		unset($cacheControlHeaders, $header, $value);
+		if (!empty($cacheControlHeaders)) {
+			HTTPCacheControl::inst()->setDirectivesFromArray($cacheControlHeaders);
+		}
 
-		if(self::$modification_date && $cacheAge > 0) {
+		if (self::$modification_date) {
 			$responseHeaders["Last-Modified"] = self::gmt_date(self::$modification_date);
+		}
+
+		// if we can store the cache responses we should generate and send etags
+		if (!HTTPCacheControl::inst()->hasDirective('no-store')) {
 
-			// Chrome ignores Varies when redirecting back (http://code.google.com/p/chromium/issues/detail?id=79758)
-			// which means that if you log out, you get redirected back to a page which Chrome then checks against
-			// last-modified (which passes, getting a 304)
-			// when it shouldn't be trying to use that page at all because it's the "logged in" version.
-			// By also using and etag that includes both the modification date and all the varies
-			// values which we also check against we can catch this and not return a 304
-			$etagParts = array(self::$modification_date, serialize($_COOKIE));
-			$etagParts[] = Director::is_https() ? 'https' : 'http';
-			if (isset($_SERVER['HTTP_USER_AGENT'])) $etagParts[] = $_SERVER['HTTP_USER_AGENT'];
-			if (isset($_SERVER['HTTP_ACCEPT'])) $etagParts[] = $_SERVER['HTTP_ACCEPT'];
-
-			$etag = sha1(implode(':', $etagParts));
-			$responseHeaders["ETag"] = $etag;
-
-			// 304 response detection
-			if(isset($_SERVER['HTTP_IF_MODIFIED_SINCE'])) {
-				$ifModifiedSince = strtotime(stripslashes($_SERVER['HTTP_IF_MODIFIED_SINCE']));
-
-				// As above, only 304 if the last request had all the same varies values
-				// (or the etag isn't passed as part of the request - but with chrome it always is)
-				$matchesEtag = !isset($_SERVER['HTTP_IF_NONE_MATCH']) || $_SERVER['HTTP_IF_NONE_MATCH'] == $etag;
-
-				if($ifModifiedSince >= self::$modification_date && $matchesEtag) {
-					if($body) {
-						$body->setStatusCode(304);
-						$body->setBody('');
-					} else {
-						header('HTTP/1.0 304 Not Modified');
-						die();
+			$etag = static::generateETag($body);
+
+			if ($etag) {
+				$responseHeaders['ETag'] = $etag;
+
+				// 304 response detection
+				if (isset($_SERVER['HTTP_IF_MODIFIED_SINCE'])) {
+					$matchesEtag = !isset($_SERVER['HTTP_IF_NONE_MATCH']) || $_SERVER['HTTP_IF_NONE_MATCH'] == $etag;
+
+					if ($matchesEtag) {
+						if ($body) {
+							$body->setStatusCode(304);
+							$body->setBody('');
+						} else {
+							// this is wrong, we need to send the same vary headers and so on
+							header('HTTP/1.0 304 Not Modified');
```

Comment:
I don't understand why this logic branch exists. Why does the presence or lack of a response body influence the way the response code is sent?

Reply:
This is pre-existing behaviour of HTTP::add_cache_headers() that should be deprecated in 4 but we still need to support in SS3:

You can call it without an SS_HTTPResponse object and it will instead perform direct `header()` calls. Yuck, but it's used in RSSFeed.php in SS3 and in principle could be used by project code.

Reply:
Oh right, it's checking for the presence of a response object implicitly by checking for `$body`

Reply:
RSSFeed could still take a response here, and we could make the missing response arg a deprecation notice.